ledger-api/grpc-definitions/com/daml/ledger/api/v1/admin/user_management_service.proto

```diff
@@ -160,6 +184,29 @@ message GetUserResponse {
     User user = 1;
 }
 
+
+
+// Required authorization: ``HasRight(ParticipantAdmin)``
+// NOTE: The authenticated participant user is not allowed to deactivate itself.
+message UpdateUserRequest {
+    // The user to update.
+    // Modifiable fields:
+    // 1. 'primary_party',
+    // 2. 'is_deactivated',
+    // 3. 'metadata' (with the 'replace-all' semantics).
```

Comment:
May also be nice to include `rights` here with similar 'replace-all' semantics.

Reply:
Not sure. Currently we have a low fixed limit on the number of rights (1k ?) per user, but we were thinking about making it essentially unlimited in the future.
If we were do the replace-all semantics it might get problematic in practice with very large payloads?

Reply:
Won't that also be a problem with respect to `CreateUser`, which similarly allows for the specification of a set of initial rights at the time of user creation?

Reply:
I was thinking about it on two levels:
1. Whether we want/need to track concurrent changes to user's rights. I think is not yet ( but maybe in the future we'd like to do that).
2.  Whether rights are part of the User resource. Conceptually it'd make sense however current api design suggest otherwise as rights are not part of the User message.
message CreateUserRequest {
    // The user to create.
    // Required
    User user = 1;

    // The rights to be assigned to the user upon creation,
    // which SHOULD include appropriate rights for the ``user.primary_party``.
    // Required
    repeated Right rights = 2;
}

So I'm planning to treat user rights as separate from user resource itself and leave updating them exclusively to the existing GrantRights/RevokeRight apis, and do not track changes to user rights in the user's resourceVersion.

Reply:
If you think about a consumer that's interested in remaining up to date on a `User`s rights then having them included in the resourceVersion would be really nice, wouldn't it? and it seems that we wouldn't expect too much contention coming from concurrent rights and update user changes, so it should be OK for the resourceVersion to update a bit more frequently.

Moreover, this would allow the right's tracking code for streams in the Ledger API server to just track the users' resource version, which is much cheaper than the complete set of rights.

What do you think?

Reply:
Is your intention to only extend the user's resourceVersion to be able to track changes to the user's rights? Or does it also include exposing means for concurrent change control on `GrantRights`/`RevokeRights` rpcs?

The latter could be implemented by adding `ObjectMeta metadata` to respective request and response messages and, to keep things consistent, returning the full ObjectMeta (i.e. both resource_version and annotations) for `GrantRights`/`RevokeRights`responses.

Reply:
> Is your intention to only extend the user's resourceVersion to be able to track changes to the user's rights? Or does it also include exposing means for concurrent change control on `GrantRights`/`RevokeRights` rpcs?
> 
> The latter could be implemented by adding `ObjectMeta metadata` to respective request and response messages and, to keep things consistent, returning the full ObjectMeta (i.e. both resource_version and annotations) for `GrantRights`/`RevokeRights`responses.

That is a natural extension of my proposal. It's up to you and @mziolekda to decide when you'd like to do the latter. I'd just suggest to track all mutations, as that's the basis for this.

Reply:
I'm ok with limiting it only to tracking user-right changes in user's resourceVersion, because it doesn't change the API further.

I'd leave out the complete support of concurrent change control for user-rights from this design to avoid scope creep - at least until we are confident we can deliver the required features in time.